### ISSUE TITLE
Add `namespace` option to `provideState`

### ISSUE THREAD

**Issue by divmain** [FormidableLabs/freactal#8]
When `provideState` is invoked like so:

```javascript
provideState({
  namespace: "myNamespace",
  initialState: () => ({ value: "myValue" }),
  effects: { setVal: softUpdate((state, value) => ({ value })) }
});
```

All state values and effect functions will be silo'd within that namespace.  Consuming from a child component would look like:

```javascript
injectState(({ state, effects }) => (
  <div>
    {state.myNamespace.value}
    <button
      onClick={() => effects.myNamespace.setVal("newValue")}
    >
      Click me!
    </button>
  </div>
));
```

**Comment by jimthedev** [comment]
IMO this should be the default and you should have to opt in to using in to using the default global namespace.

**Comment by fingermark** [comment]
As mentioned on gitter, I just tried implementing a form using freactal and immediately used a namespacing pattern.  It'd be nice if this was supported as an option.

**Comment by zebulonj** [comment]
Would the `effects` object exposed to effects defined on that name-spaced provider be the global `effects` object (inheriting from parents), or just the local scope? I'd venture that it should be the global `effects` object, or it would be impossible to call effects defined by parents (a fundamental benefit of the composition described in the documentation).

**Comment by divmain** [comment]
@zebulonj, I agree that it would have to be the global effects object.  Doing the other would be harder, actually.

For those that are wondering, the difficulty with implementing this is not actually with the namespaced state or effects, it is in tracking the hierarchical dependencies of namespaced state-keys for computed values.  I really want to see this feature land, but it's taking a backseat for a bit until all the kinks have been ironed out.
